### DISCUSSION TITLE
Adding getRelation and tuple iterator to SWIG Interface

### DISCUSSION THREAD

**Discussion by cwarden** [souffle-lang/souffle#2004] [Q&A]
I've got a [simple go application](https://github.com/cwarden/souffle-go-example/blob/5ddd4f499f3accc4d57340eb63c9cb6ef365bb3a/example.go) calling a souffle program using SWIG.

I'd like to update the SWIG interface so I can get a relation's tuples.  I updated the SWIG interface to [add `getRelation`](https://github.com/cwarden/souffle-go-example/commit/5f55d977d8748a1f3cd2776ab4c3b6fdf1703c26), but am not sure how to go about adding bindings for the iterator.  Instead of returning `souffle::Relation*` directly, should I wrap it in a new `SWIGSouffleRelation` class similar to `SWIGSouffleProgram`, and then implement a method to iterate using the underlying iterator?  Is there a simpler or more idiomatic solution?

Ultimately, I'm thinking the interface would allow for something like
```
	p := SwigInterface.NewInstance("example")
	p.LoadAll(".")
	p.Run()
	r := p.GetRelation("path")
	for {
		tuple, ok := r.Next()
		if !ok {
			break
		}
		fmt.Printf("%+v", tuple)
	}
```

**Comment by b-scholz** [comment]
I can vaguely remember we had an initial prototype accessing the tuples of a relation. Indeed, you will need a wrapper because we could not directly use Souffle's native C++ interface in SWIG. Great that you are looking into this issue!

**Reply by b-scholz** [reply]
I found an earlier version from Andres. Unfortunately, it seems that the relational interface got lost. Please can you have a look:
[Interfaces with examples.zip](https://github.com/souffle-lang/souffle/files/6965722/Interfaces.with.examples.zip)

**Reply by cwarden** [reply]
Thanks.  It's looking promising: https://github.com/cwarden/souffle-go-example/commit/bbeb0eaba27edc753045fc378100c3370167c615

**Reply by b-scholz** [reply]
I can remember now that students who worked on the swig interface had some issues w.r.t. relations. I cannot remember the details. 

If you have time, it would be great if you could re-introduce the relations again.  You find the swig sources of soufflé in `src/include/souffle/swig`.

**Reply by cwarden** [reply]
@b-scholz how does #2014 look?  Suggestions for improvement welcome.

Here's the example using the updated interface: https://github.com/cwarden/souffle-go-example/blob/688c8f33d5ae73cf875e52b994032d46166cac5b/example.go
